src/vidpf.rs

```diff
@@ -226,6 +232,65 @@ impl<W: VidpfValue, const NONCE_SIZE: usize> Vidpf<W, NONCE_SIZE> {
         })
     }
 
+    /// [`Vidpf::eval_with_cache`] evaluates the entire `input` and produces a share of the
+    /// input's weight. It reuses computation from previous levels available in the
+    /// cache
+    pub fn eval_with_cache(
+        &self,
+        key: &VidpfKey,
+        public: &VidpfPublicShare<W>,
+        input: &VidpfInput,
+        cache_tree: &mut BinaryTree<VidpfEvalCache<W>>,
+        nonce: &[u8; NONCE_SIZE],
+    ) -> Result<VidpfValueShare<W>, VidpfError> {
```

Comment:
This implementation looks correct, but it's a bit messy in ways that reveal some sharp edges in the `BinaryTree` API. Namely, there are a bunch of points where we unwrap and `Option`. Also, it's awkward that we have to re-construct an index each time we want to insert a node.

I think it would be cleaner to just construct the tree ourselves. Here's what I came up with: this requires some refactoring of the non-public API, but we're of course free to make changes to this as needed.

```rust
let n = input.len();
if n > public.cw.len() {
    return Err(VidpfError::InvalidAttributeLength);
}

if cache_tree.root.is_none() {
    cache_tree.root = Some(Box::new(Node::new(VidpfEvalCache {
        state: VidpfEvalState::init_from_key(key),
        share: W::zero(&self.weight_parameter), // not used
    })));
}

let mut sub_tree = cache_tree.root.as_mut().expect("root was visited");
for (level, bit) in input.iter().enumerate() {
    sub_tree = if !bit {
        if sub_tree.left.is_none() {
            let (new_state, new_share) =
                self.eval_next(key.id, public, input, level, &sub_tree.value.state, nonce)?;
            sub_tree.left = Some(Box::new(Node::new(VidpfEvalCache {
                state: new_state,
                share: new_share,
            })));
        }

        sub_tree.left.as_mut().expect("left child was visited")
    } else {
        if sub_tree.right.is_none() {
            let (new_state, new_share) =
                self.eval_next(key.id, public, input, level, &sub_tree.value.state, nonce)?;
            sub_tree.right = Some(Box::new(Node::new(VidpfEvalCache {
                state: new_state,
                share: new_share,
            })));
        }

        sub_tree.right.as_mut().expect("right child was visited")
    }
}

Ok(sub_tree.value.to_share())
```